src/pages/Groups.tsx

```diff
@@ -0,0 +1,253 @@
+
+const all = "Tutti"
+
+const deltaTime = 500 //ms
+
+export const Groups: MainStackScreen<"Groups"> = () => {
+    //keep track of latest search request time (successful or not)
+    const [lastSearchTime, setLastSearchTime] = useState<undefined | Date>(
+        undefined
+    )
+    //for forcing api search request if needeed
+    const [needSearching, setNeedSearching] = useState<boolean>(false)
+
+    //for triggering api search request side effect (used in conjunction with `needSearching`)
+    //reason : I want to trigger it only if I set needSearching to True, not to False
+    const [rescheduleSearch, setRescheduleSearch] = useState<number>(0)
+
+    //for triggering api request when user doesnt write in a time span of `deltaTime`
+    //reason: I don't want to send api request on every character change.
+    const [prevSearch, setPrevSearch] = useState("")
+
+    const [search, setSearch] = useState("")
+
+    const [isSearching, setIsSearching] = useState(false)
+
+    // ? I'd like to typecheck year, course, etc... but I need dynamic type checking
+    // ? and it's very ugly, and need to use type guards or some library, so maybe it is not worth it?
+    const [year, setYear] = useState<string>(all)
+
+    const [course, setCourse] = useState<string>(all)
+
+    const [type, setType] = useState<string>(all)
+
+    const [platform, setPlatform] = useState<string>(all)
+
+    const [groups, setGroups] = useState<Group[] | undefined>(undefined)
+
+    //actually displayed groups (Ordered by year)
+    const [orderedGroups, setOrderedGroups] = useState<Group[] | undefined>(
+        undefined
+    )
+
+    const [isModalShowing, setIsModalShowing] = useState(false)
+
+    const [modalGroup, setModalGroup] = useState<Group | undefined>(undefined)
+
+    //tracking first render
+    const isMounted = useMounted()
+
+    /**
+     * Api search request side effect.
+     *
+     * How this works (general ideas):
+     * 1) gets called every time search text changes or rescheduleSearch is incremented (for forcing search request)
+     * 2) if len(search) < 5 do nothing and delete stored groups if there are any.
+     * 3) if no other searches were done before, send search request to api. Store time of search in `lastSearchTime`.
+     * 4) successive requests are allowed if last search request time (successful or not) was more than `deltaTime` (500ms) ago
+     * 5) if a request is not allowed, store `lastTimeSearch` anyway, then check `deltaTime` (500ms) later if the user has kept writing.
+     * if he has, then do nothing because side effect will be called again, if he hasn't,
+     * then force search request with latest search value.
+     */
+    const searchGroups = async () => {
+        if (isMounted) {
+            if (search.length < 5) {
+                if (groups !== undefined) {
+                    setGroups(undefined)
+                }
+                return
+            }
```

Comment:
veramente vogliamo avere 5 come minimo della ricerca? per trovare un gruppo di informatica devi scrivere almeno "infor" quando molta gente già lo abbrevia a "info", sinceramente non è che dobbiamo preoccuparci troppo della banda consumata, sono poi json mandati avanti e indietro, se proprio vuoi mettere un minimo anche solo 2 basta, ma secondo me nemmeno serve

Reply:
Okay, allora lo tolgo del tutto

Reply:
Sto riscontrando dei seri problemi di performance quando vengono fatte richieste in cui partono delle chiamate intermedie di tipo 1-2-3 caratteri. Sembra che il timer si fermi finchè tutta la scrollView non viene renderizzata, ho provato togliendo gli svg ma il problema rimane. Se tipo cerco Informatica e parte una richiesta intermedia con solo "I", arriva una risposta di 1200 gruppi e la risposta di "Informatica" viene mostrata a schermo dopo tipo 10 secondi. Io metterei come minimo di ricerca 4 caratteri per evitare questo tipo di problemi, altrimenti potrei provare a utilizzare una scroll view con lazy loading (?).

Reply:
ah shit, effettivamente "sti cazzi della banda" è divertente finché non causa 1400 rendering che bloccano il main thread lol
io direi che effettivamente, a prescindere, qua dovremmo usare una [FlatList](https://reactnative.dev/docs/flatlist) al posto di una ScrollView che dovrebbe essere più performante essendo che sono sempre gli stessi componenti ripetuti
poi magari sarebbe anche sensato troncare la lunghezza della risposta e renderizzare solo i primi n elementi (con n da definirsi), nessuno starà mai lì a scorrere centinaia di gruppi nella piccola schermata dell'app spero ahah
e si magari fare la ricerca con un solo carattere è un po' esagerato ahah

Reply:
Wow sembra molto molto più veloce con la flatlist! Ottimo.

Abbasso il limite a 3 caratteri ok?